toolchain/check/handle_impl.cpp

```diff
@@ -213,39 +212,15 @@ static auto PopImplIntroducerAndParamsAsNameComponent(
   Parse::NodeId first_param_node_id =
       context.node_stack().PopForSoloNodeId<Parse::NodeKind::ImplIntroducer>();
 
-  // Subtracting 1 since we don't want to include the final `{` or `;` of the
-  // declaration when performing syntactic match.
-  auto end_node_kind = context.parse_tree().node_kind(end_of_decl_node_id);
-  CARBON_CHECK(end_node_kind == Parse::NodeKind::ImplDefinitionStart ||
-               end_node_kind == Parse::NodeKind::ImplDecl);
+  auto node_kind = context.parse_tree().node_kind(end_of_decl_node_id);
+  CARBON_CHECK(node_kind == Parse::NodeKind::ImplDefinitionStart ||
+               node_kind == Parse::NodeKind::ImplDecl ||
+               node_kind == Parse::NodeKind::WhereOperand);
   Parse::Tree::PostorderIterator last_param_iter(end_of_decl_node_id);
+  // Subtracting 1 since we don't want to include the final `{`, `;`, or `where`
+  // of the declaration when performing syntactic match.
   --last_param_iter;
```

Comment:
Move this to the caller so we can do the check that we only do the skipping the `where` nodes if the last node is a `WhereExpr`, as suggested in open discussion.
```suggestion
```

(Also change `*last_param_iter` -> `end_of_decl_node_id` below)

Reply:
Can you take a look at the current approach and let me know if you still want this moved?